variety/AddWallhavenDialog.py

```diff
@@ -0,0 +1,192 @@
+        if "apikey" in params:
+            self.ui.apikey.set_text(params["apikey"][0])
+
+    @staticmethod
+    def parse_location(location):
+        url_parts = urllib.parse.urlparse(location)
+        params = urllib.parse.parse_qs(url_parts.query)
+
+        return url_parts, params
+
+    def ok_thread(self):
+        def _start_ui():
+            self.ui.buttonbox.set_sensitive(False)
+            self.ui.message.set_visible(True)
+            self.ui.query.set_sensitive(False)
+            self.ui.spinner.set_visible(True)
+            self.ui.spinner.start()
+            self.ui.error.set_label("")
+            self.ui.use_apikey_or_not.set_sensitive(False)
+            self.ui.apikey.set_sensitive(False)
+
+        Util.add_mainloop_task(_start_ui)
+
+        query = self.ui.query.get_text().strip()
+        query = WallhavenDownloader(None, query).web_url
+        usekey = self.ui.use_apikey_or_not.get_active()
+
+        valition_empty_apikey = False
+        invalid_msg = None
+        if usekey:
+            apikey = self.ui.apikey.get_text().strip()
+            if apikey is None or apikey == "":
+                valition_empty_apikey = True
+                invalid_msg = "Please type your apikey."
+            else:
+                # Append apikey to the original url, so that we can access those NSFW images or other settings
+                if "&apikey=" not in query:
+                    query = query + "&apikey=" + apikey
+                if "&usekey=" not in query:
+                    query = query + "&usekey=" + str(usekey)
+
+        else:
+            url_parts, queries = self.parse_location(query)
+            query = self.clean_url(url_parts, queries)
+
+        if not valition_empty_apikey:
+            final_query, invalid_msg = self.validate(query)
+
+        def _stop_ui():
+            if invalid_msg:
+                self.ui.error.set_label(invalid_msg)
+                self.ui.spinner.stop()
+                self.ui.query.set_sensitive(True)
+                self.ui.message.set_visible(False)
+                self.ui.spinner.set_visible(False)
+                if valition_empty_apikey:
+                    self.ui.apikey.grab_focus()
+                else:
+                    self.ui.query.grab_focus()
+                self.ui.use_apikey_or_not.set_sensitive(True)
+                self.ui.apikey.set_sensitive(True)
+                self.ui.buttonbox.set_sensitive(True)
+            else:
+                self.commit(final_query)
+                self.destroy()
+
+        Util.add_mainloop_task(_stop_ui)
+
+    def clean_location(self, location):
+        """ Remove apikey and usekey from a given url """
+        url_parts, queries = self.parse_location(location)
+        return self.clean_url(url_parts, queries)
+
+    @staticmethod
+    def clean_url(url_parts, queries):
+        """ Remove apikey and usekey from a given url """
+        if len(queries) == 0:
+            path = url_parts.path
```

Comment:
We shouldn't need this low-level URL parsing - we should be able to use urllib.parse.urlparse and urllib.parse.parse_qs.

Reply:
It's tricky here.
# Explanation
Urls inputted by users is actually stored in ```~/.config/variety/variety.conf```, which are combined with other options. For example: 
```
src8 = True|wallhaven|https://wallhaven.cc/search?q=keyword&categories=001&purity=011&atleast=1920x1080&ratios=16x9&sorting=relevance&order=desc&apikey=xxxxx&usekey=True.
```

I have to remove apikey and usekey from this kind of url if users choose not to access the NSFW pictures by using apikey. 

And I type a sort of url like ```https://wallhaven.cc/tag/93679```, the final url stored in variety.conf will like ```https://wallhaven.cc/tag/93679&apikey=xxxxx&usekey=True```.

I append apikey and usekey in the end of the url and consider apikey and usekey arguments.

But urllib.parse.parse_qs is unable to extract arguments from ```https://wallhaven.cc/tag/93679&apikey=xxxxx&usekey=True``` so that I can not remove apikey and usekey if necessary.

So I write such a special function to handle this kind of case.

There are better measures or tools to handle this situation somewhere, I don't know yet...

# Test
```Python
def printArgsOfGivenUrl(location):
    import urllib
    unquote_location = urllib.parse.unquote(location, ```UTF-8```)
    url_parts = urllib.parse.urlparse(unquote_location)
    args = urllib.parse.parse_qs(url_parts.query, keep_blank_values=True)
    print(args)
```
## Test 1
```Python
printArgsOfGivenUrl("https://wallhaven.cc/search?q=keyword&categories=001&purity=011&atleast=1920x1080&ratios=16x9&sorting=relevance&order=desc&apikey=xxxxx&usekey=True")
```
Output:
```
{'q': ['keyword'], 'categories': ['001'], 'purity': ['011'], 'atleast': ['1920x1080'], 'ratios': ['16x9'], 'sorting': ['relevance'], 'order': ['desc'], 'apikey': ['xxxxx'], 'usekey': ['True']}
```
## Test 2
```
printArgsOfGivenUrl("https://wallhaven.cc/tag/93679&apikey=xxxxx&usekey=True")
```
Output: 
```
Empty
```

Reply:
If we persist  apikey in other places, perhaps there's no need to have this special function at all.